### ISSUE TITLE
unique registration_id as textfield broken on mysql

### ISSUE THREAD

**Issue by henriklindgren** [xtrinch/fcm-django#236]
Since version 1.0.13

Probably this commit (registration_id unique) https://github.com/xtrinch/fcm-django/commit/f19227f9468cbd34f45513d5b7f12e64e9e93504

Error message
```
django.db.utils.OperationalError: (1170, "BLOB/TEXT column 'registration_id' used in key specification without a key length")
```

This seems to be a Django <-> MySQL issue with textfields. To have it work with the unique flag it needs to be a charfield, without the unique flag it works as a textfield.

**Comment by henriklindgren** [comment]
I've had a look at this. From a database agnostic view it would be better to use a CharField because we can add unique to it and not care about which database it runs on. For FCM registration_ids a better fit is TextField because of the quote below from Google engineer Mr Johns. 

Trevor Johns is likely someone to follow the advice from:

> The documentation doesn't specify any pattern, therefore any valid string is allowed. The format may change in the future; please do not validate this input against any pattern, as this may cause your app to break if this happens.
> 
> As with the "registration_id" field, the upper bound on size is the max size for a cookie, which is 4K (4096 bytes).
>  https://stackoverflow.com/a/12502351 

Now, if we opt for using CharField, this would require an explicit rollback of the migrations to pre-unique, then remove old migration path and apply an alter migration. This is tedious as hell for end users of a library.

@xtrinch I'm open for any ideas and can write the change once your happy with the path to take.

**Comment by xtrinch** [comment]
@henriklindgren We need solve this in a backward compatible way. We can modify the existing migration to do nothing on mysql, and in the new migration just simply alter the field to CharField with a unique index.

What do you think?
I found some insight here https://stackoverflow.com/questions/32384547/django-migration-runsql-conditional-on-database-type

**Comment by merwok** [comment]
A migration that’s part of a released version should not be changed, or it will cause trouble and require manual steps for people who are already on that version.

Let’s call the existing migration that doesn’t work on mysql M1.
What you could do is change the model to generate a migration M2 that undoes the changes, change the model to change the field in the way that achieves all goals (size and uniqueness) and get a migration M3 for that, then create a squashed migration M4 that’s equivalent to M1-M2-M3.  People already on M1 will be able to run the intermediate migrations, and people who have the previous version will migrate directly to M4.

**Comment by xtrinch** [comment]
> A migration that’s part of a released version should not be changed, or it will cause trouble and require manual steps for people who are already on that version.
> 
> Let’s call the existing migration that doesn’t work on mysql M1.
> What you could do is change the model to generate a migration M2 that undoes the changes, change the model to change the field in the way that achieves all goals (size and uniqueness) and get a migration M3 for that, then create a squashed migration M4 that’s equivalent to M1-M2-M3.  People already on M1 will be able to run the intermediate migrations, and people who have the previous version will migrate directly to M4.

It cant cause problems if all we do is skip the existing migrations content for mysql, as mysql users cannot run it.

There also does not exist an alternative that covers the size and uniqueness constraints for mysql. Char field is too small and text field cannot be made unique.

So the best option is to not penalize other users for mysql's + django lack of support for this and enable the uniqueness of the text field to work for everybody else.

**Comment by xtrinch** [comment]
If you could add some tests, that would be 10/10.

Also, you don't really need to do any checking, this package will update the DB entry upon duplicate registration id (see https://github.com/xtrinch/fcm-django/blob/5bbd7d0168ab26221621d6749522b35e0581f56d/fcm_django/api/rest_framework.py#L111). So basically it already does the manual checking and this isn't such a horrible overhead.

The unique index is more of an optimization, so it's fine if for mysql it isn't there at all.

**Comment by henriklindgren** [comment]
Having trouble finding the time to test it, maybe later this evening, this is the WIP. 

https://github.com/xtrinch/fcm-django/compare/master...henriklindgren:fcm-django:bugfix/issue236

The idea was to skip database migration state, but keep app state moving forward. From what I've understood the unique flag should not cause issues even if we skip it, but I need to verify it all. Soon come :)

**Comment by xtrinch** [comment]
Awesome!
@henriklindgren It might generate migrations when mysql users run `makemigrations`, this needs to be somehow avoided if yes

**Comment by henriklindgren** [comment]
Tested migration behaviour, makemigrations is not triggered, wrote specifics on what's tested at the moment in the PR https://github.com/xtrinch/fcm-django/pull/238

**Comment by xtrinch** [comment]
Fixed in 1.0.15

**Comment by croozeus** [comment]
Hey there! I see this was fixed last month but I encountered this issue when I derived my model from AbstractFCMDevice and ran migrations. 

I'm using MySQL. Any ideas on how I can work around this? Thanks!
